src/rust/engine/workunit_store/src/lib.rs

```diff
@@ -701,6 +706,53 @@ impl WorkunitStore {
         });
     }
   }
+
+  ///
+  /// Records an observation of a time-like metric into a histogram.
+  ///
+  pub fn record_observation(&self, metric: ObservationMetric, value: u64) {
+    let mut histograms_by_metric = self.observation_data.observations.lock();
+    histograms_by_metric
+      .entry(metric)
+      .and_modify(|h| {
+        let _ = h.record(value);
+      })
+      .or_insert_with(|| {
+        let mut h = hdrhistogram::Histogram::<u64>::new(3).expect("Failed to allocate histogram");
+        let _ = h.record(value);
+        h
+      });
+  }
+
+  ///
+  /// Return all observations in binary encoded format.
+  ///
```

Comment:
Rather than sending the entire histogram, you might just send a few percentiles: `p0, p50, p90, p95, p99, p100` etc.

Reply:
This PR technically doesn't send anything. That would be done by the buildsense plugin. This PR would need support for extracting the percentiles though to support that downstream change. That said, the buildsense plugin could use the Python version of `HdrHistogram` to process this data itself.

@asherf: Thoughts?

Reply:
I prefer to process the data on the server/service side rather than the plugin.
keeps things simple and flexible.

Reply:
> This PR technically doesn't send anything. That would be done by the buildsense plugin.

I wouldn't even expose the entire histogram to the `StreamingWorkunitHandler`... could provide flexibility by exposing the percentiles to return, maybe.

I think that unless you are planning to aggregate multiple histograms on the server, there just isn't a good reason (or easy way) to send and then decode the whole thing.

Reply:
I prefer that we keep Pants and the Buildsense plugin simple to start. If it results in too much data, we can scale it back to just percentiles in a subsequent revision. The "smarts" for processing the observation histograms should be kept in the Buildsense server.

Reply:
> I think that unless you are planning to aggregate multiple histograms on the server, there just isn't a good reason (or easy way) to send and then decode the whole thing.

And I may want to aggregate in exactly that way. It'll make for nice dashboards.

Reply:
> I prefer that we keep Pants and the TC plugin simple to start.

AFAICT, percentiles are _much_ simpler than sending a binary histogram to the server, figuring out how to decode them in a remote service (which might not have embedded rust or java, making it challenging to deserialize these), adding a "binary histogram" storage strategy alongside other metrics storage (which tends to be string key to integer value), etc.

> And I may want to aggregate in exactly that way. It'll make for informative dashboards.

Binary histograms are a fairly large commitment, not just in the client, but also in the server. Would want to make sure that everyone is on board with that.

Reply:
>  figuring out how to decode them in a remote service (which might not have embedded rust or java, making it challenging to deserialize these), adding a "binary histogram" storage strategy alongside other metrics storage (which tends to be string key to integer value), etc.

Ordinarily yes but I intentionally chose the [HDR Histogram project](https://github.com/HdrHistogram) because it is available in multiple languages including [Rust](https://github.com/HdrHistogram/HdrHistogram_rust) and [Python](https://github.com/HdrHistogram/HdrHistogram_py). Our server code can just consume the histogram using the Python library.

Reply:
Ok. Would just be very sure that your server is ready to store them.

Reply:
Agreed. It will be a discussion that I'll have with @asherf.